### PR TITLE
feat: GuessLex: print inferred termination argument

### PR DESCRIPTION
**Pull request by nomeata**
With

    set_option showInferredTerminationBy true

this prints a message like

    Inferred termination argument:
    termination_by
      ackermann n m => (sizeOf n, sizeOf m)

it tries hard to use names that

 * match the names that the user used, if present
 * have no daggers (so that it can be copied)
 * do not shadow each other
 * do not shadow anything from the environment (just to be nice)

it does so by appending sufficient `'` to the name.

Some of the emitted `sizeOf` calls are unnecessary, but they are needed
sometimes with dependent parameters. A follow-up PR will not emit them
for non-dependent arguments, so that in most cases the output is pretty.

Somewhen down the road we also want a code action, maybe triggered by
`termination_by?`. This should come after #2921, as that simplifies that
feature (no need to merge termination arguments from different cliques
for example.)

### CODE REVIEW

**Comment by nomeata** [comment]
This seems to work, but there may be room for refinements:

* Is this a good name for the option?

* Is the way I generate fresh, accessible names appropriate?

* Why does 
  ```
  let fn : Ident := mkIdent (← unresolveNameGlobal element.declName)
  ```
  not remove the local namespace in
  ```
  termination_by
    VarNames.shadow1 x2 x2 => sizeOf x2
  ```

**Comment by nomeata** [comment]
Ok, figued out the last question myself :-)

File: src/Lean/Elab/PreDefinition/WF/GuessLex.lean [hunk 1]
```diff
@@ -535,22 +551,34 @@ def buildTermWF (declNames : Array Name) (varNamess : Array (Array Name))
...
-    trace[Elab.definition.wf] "Using termination {declName}: {vars} => {body}"
     termByElements := termByElements.push
       { ref := .missing
         declName, vars, body,
         implicit := true
       }
   return .ext termByElements
 
+open Parser.Command in
+def delabTermWF (preDefs : Array PreDefinition) : TerminationWF → MetaM Syntax
```
Comment (nomeata) [inline src/Lean/Elab/PreDefinition/WF/GuessLex.lean:570]: TODO: move next to the corresponding parser

**Comment by nomeata** [comment]
I built mathlib with the option set to `true`, these the termination metrics that are printed:

<details>

```
termination_by
  aux i j H b => sizeOf i
termination_by
  SatisfiesM_foldlM.go i j b h₁ h₂ H => sizeOf i
termination_by
  takeWhile₂_eq_takeWhile₂TR.go as bs acca accb => sizeOf as
termination_by
  zoom_ins path t' path' t cmp => sizeOf t
termination_by
  zoom_del cut path t' path' t => sizeOf t
termination_by
  firstDiffPos_loop_eq l₁ l₂ r₁ r₂ stop p hl₁ hl₂ hstop => sizeOf r₁
termination_by
  splitAux_of_valid l m r acc => sizeOf r
termination_by
  succ_div x1 x2 => sizeOf x1
termination_by
  mem_of_subset' x1 x2 x3 x4 => sizeOf x1
termination_by
  maxPowDiv.go k p n => sizeOf n
termination_by
  go_succ k p n => sizeOf n
termination_by
  xgcdAux x1 x2 x3 x4 x5 x6 => sizeOf x1
termination_by
  decidableEq x1 x2 => sizeOf x1
termination_by
  interleave x1 x2 => sizeOf x1
termination_by
  append_bind xs ys f => sizeOf xs
termination_by
  ltb s₁ s₂ => sizeOf s₁
termination_by
  inductionOn it₁ it₂ ind eq base₁ base₂ => sizeOf it₁
termination_by
  cprankMax_add x1 x2 x3 x4 x5 x6 => sizeOf x1
termination_by
  exists_subgroup_card_pow_prime_le x1 x2 x3 => sizeOf x2
```

</details>

Only 19 definitions in all of mathlib need the decreasing argument to be guessed (all others are either structural, have an explicit `terminating_by` or have only one plausible argument). That’ll make me write more tests :-)

None of these use more than one argument, which is expected, as until very recently, a definition that needed a lexicographic argument would not be accepted.

There are 33 `termination_by` annotations that were necessary before, and should now no longer be necessary:

<details>

```
Mathlib/Computability/Ackermann.lean:  termination_by ack m n => (m, n)
Mathlib/Computability/Ackermann.lean:  termination_by ack_strictMono_right m x y h => (m, x)
Mathlib/Computability/Ackermann.lean:  termination_by add_lt_ack m n => (m, n)
Mathlib/Computability/Ackermann.lean:  termination_by ack_strict_mono_left' x y => (x, y)
Mathlib/Computability/PartrecCode.lean:  termination_by evaln k c => (k, c)
Mathlib/Data/List/Infix.lean:termination_by decidableSuffix l₁ l₂ => (l₁, l₂)
Mathlib/Data/List/Infix.lean:termination_by decidableInfix l₁ l₂ => (l₁, l₂)
Mathlib/Data/Nat/Choose/Basic.lean:  termination_by multichoose a b => (a, b)
Mathlib/Data/Nat/Hyperoperation.lean:  termination_by hyperoperation a b c => (a, b, c)
Mathlib/RingTheory/Polynomial/Hermite/Basic.lean:termination_by _ n k => (n, k)
Mathlib/SetTheory/Game/Basic.lean:  termination_by _ => (x, y)
Mathlib/SetTheory/Game/Basic.lean:  termination_by _ => (x, y)
Mathlib/SetTheory/Game/Basic.lean:  termination_by _ => (x, y, z)
Mathlib/SetTheory/Game/Basic.lean:  termination_by _ => (x, y, z)
Mathlib/SetTheory/Game/Birthday.lean:termination_by birthday_congr x y _ => (x, y)
Mathlib/SetTheory/Game/Birthday.lean:termination_by birthday_add a b => (a, b)
Mathlib/SetTheory/Game/Impartial.lean:termination_by _ G H => (G, H)
Mathlib/SetTheory/Game/Impartial.lean:termination_by _ G H _ _ => (G, H)
Mathlib/SetTheory/Game/Ordinal.lean:termination_by toPGame_add a b => (a, b)
Mathlib/SetTheory/Game/PGame.lean:termination_by _ x y => (x, y)
Mathlib/SetTheory/Game/PGame.lean:termination_by _ w x y z _ _ => (x, z)
Mathlib/SetTheory/Game/PGame.lean:termination_by _ x y => (x, y)
Mathlib/SetTheory/Game/PGame.lean:termination_by _ x y => (x, y)
Mathlib/SetTheory/Game/PGame.lean:termination_by _ x y z => (x, y, z)
Mathlib/SetTheory/Game/PGame.lean:termination_by _ x y z => (x, y, z)
Mathlib/SetTheory/Ordinal/NaturalOps.lean:  termination_by nadd o₁ o₂ => (o₁, o₂)
Mathlib/SetTheory/Ordinal/NaturalOps.lean:termination_by nmul a b => (a, b)
Mathlib/SetTheory/Ordinal/NaturalOps.lean:  termination_by nadd_comm a b => (a,b)
Mathlib/SetTheory/Ordinal/NaturalOps.lean:termination_by _ => (a, b, c)
Mathlib/SetTheory/Ordinal/NaturalOps.lean:termination_by nmul_comm a b => (a, b)
Mathlib/SetTheory/Ordinal/NaturalOps.lean:termination_by nmul_nadd a b c => (a, b, c)
Mathlib/SetTheory/Ordinal/NaturalOps.lean:termination_by nmul_assoc a b c => (a, b, c)
Mathlib/SetTheory/Surreal/Basic.lean:termination_by _ x y _ _ => (x, y) -- Porting note: Added `termination_by`
```

</details>

Some of them say
```
--Porting note: `termination_by` required here where it wasn't before
```
– was lean3 able to guess lexicographic orders?

**Comment by digama0** [comment]
> Some of them say
> 
> ```
> --Porting note: `termination_by` required here where it wasn't before
> ```
> 
> – was lean3 able to guess lexicographic orders?

Actually this is a source of confusion for me: What is the default termination metric if not lexicographic? In lean 3 this was what you get out of the box, and I haven't looked at this in detail but I don't see what else one could even do here. I don't see why "guessing" is necessary and this choice of terminology suggests that I don't understand what you are trying to do.

**Comment by digama0** [comment]
lean 3 would always do `termination_by m n => (m, n)`

**Pull request merged**